tests/std/tests/P0674R1_make_shared_for_arrays/test.cpp

```diff
@@ -30,18 +289,262 @@ struct ConstructConstrainingAllocator {
+        assert_shared_use_get(p0);
+        for (int i = 0; i < 100; ++i) {
+            assert(p0[i].empty());
+        }
+    }
+    assert_construct_destruct_equal();
+
+    CustomAlloc<InitialValue> a1{};
+    {
+        shared_ptr<InitialValue[][8][9]> p1 = allocate_shared_assert<InitialValue[][8][9]>(144, a1, 2u);
+        for (int i = 0; i < 2; ++i) {
+            for (int j = 0; j < 8; ++j) {
+                for (int k = 0; k < 9; ++k) {
+                    assert(p1[i][j][k].value == 106);
+                }
+            }
+        }
+    }
+    assert_construct_destruct_equal();
+
+    {
+        shared_ptr<string[][2]> p2 = allocate_shared<string[][2]>(a0, 10, {"Meow!", "Purr"});
+        assert_shared_use_get(p2);
+        for (int i = 0; i < 10; ++i) {
+            assert(p2[i][0].compare("Meow!") == 0);
+            assert(p2[i][1].compare("Purr") == 0);
+        }
+    }
+    assert_construct_destruct_equal();
+
+    CustomAlloc<vector<int>> a3{};
+    {
+        shared_ptr<vector<int>[]> p3 = allocate_shared<vector<int>[]>(a3, 3, {9, 9, 9});
+        assert_shared_use_get(p3);
+        for (int i = 0; i < 3; ++i) {
+            for (const auto& val : p3[i]) {
+                assert(val == 9);
+            }
+        }
+    }
+    assert_construct_destruct_equal();
+
+    CustomAlloc<ThreeIntWrap> a4{};
+    {
+        shared_ptr<ThreeIntWrap[]> p4 = allocate_shared_init_assert<ThreeIntWrap[]>(5, a4, 5, {2, 8, 9});
+        for (int i = 0; i < 5; ++i) {
+            assert(p4[i].v1 == 2 && p4[i].v2 == 8 && p4[i].v3 == 9);
+        }
+    }
+    assert_construct_destruct_equal();
+
+    CustomAlloc<int> a5{};
+    { shared_ptr<int[]> p5 = allocate_shared_assert<int[]>(0, a5, 0u); } // p5 cannot be dereferenced
+    assert_construct_destruct_equal();
+
+    {
+        shared_ptr<int[][5][6]> p6 = allocate_shared<int[][5][6]>(a5, 4u);
+        assert_shared_use_get(p6);
+        for (int i = 0; i < 4; ++i) {
+            for (int j = 0; j < 5; ++j) {
+                for (int k = 0; k < 6; ++k) {
+                    assert(p6[i][j][k] == 0);
+                }
+            }
+        }
+    }
+    assert_construct_destruct_equal();
+
+    test_allocate_init_destruct_order<ReportAddress[]>(5u); // success one dimensional
+
+    test_allocate_init_destruct_order<ReportAddress[]>(20u); // failure one dimensional
+
+    test_allocate_init_destruct_order<ReportAddress[][2][2]>(2u); // success multidimensional
+
+    test_allocate_init_destruct_order<ReportAddress[][3][3]>(3u); // failure multidimensional
+}
+
 int main() {
-    ConstructConstrainingAllocator<void, int> a{};
-    (void) allocate_shared<int>(a, 42);
-    (void) allocate_shared<const int>(a, 42);
+    test_make_shared_not_array();
+    test_make_shared_array_known_bounds();
+    test_make_shared_array_unknown_bounds();
+
+    test_allocate_shared_not_array();
+    test_allocate_shared_array_known_bounds();
+    test_allocate_shared_array_unknown_bounds();
 }

```

Comment:
I'd like to see some coverage for unbounded arrays with new-extended alignments.

Reply:
Added `alignas(32)` test coverage.